src/collection/historical_signatures/signature_collector.py

```diff
@@ -0,0 +1,215 @@
+    2) Collects 1000 tx signatures starting with `tx_0`.
+    3) Collected transactions belong to sequence of blocks from `t_0` to `t_0 - k`.
+    Store to tx_signatures table all transactions from blocks `t_0 - 1` to `t_0 - k + 1` with flag `from signatures`.
+    If signature already exists - change flag to `from_signatures`.
+    Transactions from block `t_0` are handled in current data collector.
+    4) we get presumably the earliest transaction signature from `t_0 - k + 1` block and collect next 1000 tx.
+    5) repeat 3-4 until reaching the first transaction.
+    6) As first (earliest) transaction is reached - consider historical data (signatures only) for given PPK
+    at time t_0 collected.
+    If being restarted - fetch block number from last tx signature for given PPK and flag `signature`.
+    Use this block instead of `t_0`
+"""
+import logging
+import os
+import time
+import traceback
+from typing import List
+
+from solders.rpc.responses import RpcConfirmedTransactionStatusWithSignature
+from solders.signature import Signature
+from solders.transaction_status import TransactionErrorFieldless
+from solana.exceptions import SolanaRpcException
+import solana.rpc.api
+
+from src.collection.shared.generic_collector import GenericSolanaConnector, log_performance_time
+import db
+
+
+LOGGER = logging.getLogger(__name__)
+
+TX_BATCH_SIZE = 1000
+
+
+class SignatureCollector(GenericSolanaConnector):
+    def __init__(self):
+        super().__init__()
+        self.protocol = os.getenv("PROTOCOL_PUBLIC_KEY", "")
+        self._oldest_signature: Signature | None = None  # The oldest signature from the oldest completed block
+        self._signatures_completed: bool = False
+        self._oldest_completed_slot: int = 0
+        self._collected_signatures: List[RpcConfirmedTransactionStatusWithSignature] = []
+
+    @property
+    def _collection_completed(self) -> bool:
+        """
+        Flag to mark completion of collection.
+        """
+        return self._signatures_completed
+
+    @log_performance_time(LOGGER)
+    def _get_assignment(self) -> None:
+        """
+        Obtain assignment for data collection.
+        """
+        # if oldest obtained signature is defined, proceed to collection
+        if self._oldest_signature:
+            return
+        # otherwise, get retrieve oldest signature from database.
+        with db.get_db_session() as session:
+            LOGGER.info("Getting the oldest stored signature for protocol = {}.".format(self.protocol))
+            # Get the last signature collector recorded by `SignatureCollector` for given protocol
+            oldest_signature = session.query(db.TransactionStatusWithSignature.signature) \
+                .filter(db.TransactionStatusWithSignature.source == self.protocol) \
+                .filter(db.TransactionStatusWithSignature.collection_stream == db.CollectionStreamTypes.SIGNATURE) \
+                .order_by(db.TransactionStatusWithSignature.id.desc()) \
```

Comment:
Can we really count on `id` being monotonous? Shouldn't we use `block_time` or better `slot` instead?

Reply:
its primary key so it is for sure strictly monotonous, while block time and slot are not 'strictly'.

Reply:
It's monotonous in the order in which the transactions are saved to the db, but what if we save them in a different order? I.e. can it happen that an older transaction (e.g. from block 10) has a higher id than a newer transaction (e.g. from block 20)? I think so, just because we run current and historical transaction collections in parallel. Right?